streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java

```diff
@@ -97,6 +95,27 @@ public KTableKTableJoinMerger<K, VR> joinMerger() {
         return (KTableKTableJoinMerger<K, VR>) kChangeProcessorSupplier;
     }
 
+    @Override
+    public void enableVersionedSemantics(final boolean useVersionedSemantics, final String parentNodeName) {
+        enableVersionedSemantics(thisProcessorParameters(), useVersionedSemantics, parentNodeName);
+        enableVersionedSemantics(otherProcessorParameters(), useVersionedSemantics, parentNodeName);
+    }
+
+    @SuppressWarnings("unchecked")
+    private void enableVersionedSemantics(final ProcessorParameters<K, ?, ?, ?> processorParameters,
+                                          final boolean useVersionedSemantics,
+                                          final String parentNodeName) {
+        final ProcessorSupplier<K, ?, ?, ?> processorSupplier = processorParameters.processorSupplier();
+        if (!(processorSupplier instanceof KTableKTableAbstractJoin)) {
+            throw new IllegalStateException("Unexpected processor type for table-table join: " + processorSupplier.getClass().getName());
+        }
+        final KTableKTableAbstractJoin<K, ?, ?, ?> tableJoin = (KTableKTableAbstractJoin<K, ?, ?, ?>) processorSupplier;
+
+        if (parentNodeName.equals(tableJoin.joinThisParentNodeName())) {
```

Comment:
Not sure if I understand this condition? Can you elaborate? It seems to be the only place when we sue the newly added `parentNodeName` -- why do we not use it elsewhere?

Reply:
Good question -- the answer is that this join node has multiple parents, while all the other `VersionedSemanticsGraphNode` implementations only have a single parent. Because this join node has multiple parents and writes two separate processors (one for each side of the join) to the topology, when enabling versioned semantics we need a way to distinguish which side of the join we're enabling versioned semantics for. In other words, it's possible that only one side of the join is "versioned," in which case one of the two join processors should have versioned semantics enabled while the other should not. The way that we determine which side of the join to enable versioned semantics for is based on the parent node name; the processor whose "joinThis" is the parent node which has been identified as versioned is the processor for which versioned semantics will be enabled. (In the case of a self-join, both processors will satisfy this check, and both processors will have versioned semantics enabled.)

For all other `VersionedSemanticsGraphNode` implementations which only have a single parent, we could also perform an analogous parent node name check if we wanted to, but the parent node name should always match so it'd be redundant.